rust/connlib/shared/src/callbacks.rs

```diff
@@ -39,6 +42,169 @@ impl From<Ipv6Network> for Cidrv6 {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum Status {
+    Unknown,
+    Online,
+    Offline,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ResourceDescription {
+    Dns(ResourceDescriptionDns),
+    Cidr(ResourceDescriptionCidr),
+}
+
+impl ResourceDescription {
+    pub fn with_status(
+        r: crate::messages::client::ResourceDescription,
+        status: Status,
+    ) -> ResourceDescription {
+        match r {
+            crate::messages::client::ResourceDescription::Dns(r) => {
+                ResourceDescription::Dns(ResourceDescriptionDns::with_status(r, status))
+            }
+            crate::messages::client::ResourceDescription::Cidr(r) => {
+                ResourceDescription::Cidr(ResourceDescriptionCidr::with_status(r, status))
+            }
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        match self {
+            ResourceDescription::Dns(r) => &r.name,
+            ResourceDescription::Cidr(r) => &r.name,
+        }
+    }
+
+    pub fn status(&self) -> Status {
+        match self {
+            ResourceDescription::Dns(r) => r.status,
+            ResourceDescription::Cidr(r) => r.status,
+        }
+    }
+
+    pub fn id(&self) -> ResourceId {
+        match self {
+            ResourceDescription::Dns(r) => r.id,
+            ResourceDescription::Cidr(r) => r.id,
+        }
+    }
+
+    /// What the GUI clients should paste to the clipboard, e.g. `https://github.com/firezone`
+    pub fn pastable(&self) -> Cow<'_, str> {
+        match self {
+            ResourceDescription::Dns(r) => Cow::from(&r.address),
+            ResourceDescription::Cidr(r) => Cow::from(r.address.to_string()),
+        }
+    }
+}
+
+impl From<ResourceDescription> for crate::messages::client::ResourceDescription {
+    fn from(value: ResourceDescription) -> Self {
+        match value {
+            ResourceDescription::Dns(r) => {
+                crate::messages::client::ResourceDescription::Dns(r.into())
+            }
+            ResourceDescription::Cidr(r) => {
+                crate::messages::client::ResourceDescription::Cidr(r.into())
+            }
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+pub struct ResourceDescriptionDns {
+    /// Resource's id.
+    pub id: ResourceId,
+    /// Internal resource's domain name.
+    pub address: String,
+    /// Name of the resource.
+    ///
+    /// Used only for display.
+    pub name: String,
+
+    pub address_description: String,
+    pub gateway_groups: Vec<GatewayGroup>,
+
+    pub status: Status,
+}
+
+impl ResourceDescriptionDns {
+    fn with_status(
```

Comment:
I would expect a `with` function to take `self` by value. Shouldn't this be defined on the other type instead?